phasetools
================

Phasing HLA alleles with PHASE: Tools for writing input files and processing results.

### Install:

-   R packages:
    -   tidyverse
    -   argparse

### Input files

-   HLA genotypes

``` bash
head -n16 ./input/hla_genotypes.tsv
```

    subject locus   allele
    HG00096 HLA-A   A*01:01:01
    HG00096 HLA-A   A*29:02:01
    HG00096 HLA-B   B*08:01:01
    HG00096 HLA-B   B*44:03:01
    HG00096 HLA-C   C*07:01:01
    HG00096 HLA-C   C*16:01:01
    HG00096 HLA-DRB1    DRB1*03:01:01
    HG00096 HLA-DRB1    DRB1*07:01:01
    HG00097 HLA-A   A*03:01:01
    HG00097 HLA-A   A*24:02:01
    HG00097 HLA-B   B*07:02:01
    HG00097 HLA-B   B*07:02:01
    HG00097 HLA-C   C*07:02:01
    HG00097 HLA-C   C*07:02:01
    HG00097 HLA-DRB1    DRB1*13:03:01

-   Genomic start position for HLA genes (leftmost position, GRCh38 in this example).

``` bash
cat ./input/hla_coordinates.tsv
```

    locus   pos
    HLA-A   29941260
    HLA-C   31268749
    HLA-B   31353872
    HLA-DRB1    32552990

-   VCF file with known phase information (optional). The VCF needs to include at least columns "ID", "POS", and one column for each individual in the `hla_genotypes.tsv` file (order of columns is not important).

``` bash
head ./input/snps.vcf
```

    ID  POS HG00096 HG00097 HG00099
    rs1655924   29937342    1|1 0|0 0|1
    rs3823342   29945290    0|1 0|0 1|0
    rs1265081   31143898    1|0 0|0 0|1
    rs41561715  31270490    0|0 0|0 0|0
    rs2074491   31272119    0|0 0|0 0|0
    rs12199223  31274954    0|0 0|0 0|0
    rs9380240   31301055    0|1 0|0 1|0
    rs2523616   31352785    1|1 0|0 1|1
    rs75170544  32545508    0|1 0|0 0|0

### Usage

#### Write PHASE input files

``` bash
./write_phase_files.R --help
```

    usage: ./write_phase_files.R [-h] [-g HLAGENOS] [-v VCF] [-c GENECOORD]
                                 [-o OUTDIR]

    optional arguments:
      -h, --help            show this help message and exit
      -g HLAGENOS, --hlagenos HLAGENOS
                            HLA alleles for each individual
      -v VCF, --vcf VCF     VCF file with known phase variants (Optional)
      -c GENECOORD, --genecoord GENECOORD
                            HLA genes start position
      -o OUTDIR, --outdir OUTDIR
                            Output directory

##### Example

``` bash
./write_phase_files.R -g ./input/hla_genotypes.tsv -c ./input/hla_coordinates.tsv -v ./input/snps.vcf -o ./out
```

### Process phase results

``` bash
./process_phase_results.R --help
```

    usage: ./process_phase_results.R [-h] [-p PHASEOUT] [-l LOCI] [-c HLACODES]
                                     [-o OUTDIR]

    optional arguments:
      -h, --help            show this help message and exit
      -p PHASEOUT, --phaseout PHASEOUT
                            PHASE output file with phasing results
      -l LOCI, --loci LOCI  input_loci.tsv file generated by write_phase_files
      -c HLACODES, --hlacodes HLACODES
                            hla_allele_codes.txt generated by write_phase_files
      -o OUTDIR, --outdir OUTDIR
                            output directory

##### Example

``` bash
./process_phase_results.R -p ./out/phase.out -l ./out/input_loci.txt -c ./out/hla_allele_codes.txt -o ./out
```

### Output

##### 1. Phased HLA haplotypes

``` bash
cat ./out/phase_haps.tsv
```

    subject hap HLA-A   HLA-B   HLA-C   HLA-DRB1
    HG00096 1   A*01:01:01  B*08:01:01  C*07:01:01  DRB1*03:01:01
    HG00096 2   A*29:02:01  B*44:03:01  C*16:01:01  DRB1*07:01:01
    HG00097 1   A*03:01:01  B*07:02:01  C*07:02:01  DRB1*13:03:01
    HG00097 2   A*24:02:01  B*07:02:01  C*07:02:01  DRB1*15:01:01
    HG00099 1   A*01:01:01  B*08:01:01  C*07:01:01  DRB1*03:01:01
    HG00099 2   A*68:01:02  B*44:02:01:01   C*07:04:01  DRB1*11:01:01

##### 2. Inferrences with probability higher than the threshold (default p=0.9)

0 = phasing with probability &lt; 0.9

1 = phasing with probability &gt; 0.9

``` bash
cat ./out/phase_haps_conf.tsv
```

    subject HLA-A   HLA-B   HLA-C   HLA-DRB1
    HG00096 0   0   0   0
    HG00097 0   1   1   0
    HG00099 0   0   0   0
